### PR TITLE
optimize: fix up cg/bfgs initial step sizes

### PR DESCRIPTION
**Pull request by pv**
Change the initial step size in CG/BFGS so that it is |dx0| ~ 1 regardless of the scaling of the function, similarly to L-BFGS-B. Previously, it was |dx0| ~ |grad| which is just wrong.

Also, prior to Scipy 0.12.0 it used to be |dx0| ~ 1/|grad|. This also makes some sense, if you assume typical values of the function are of order 1.
- [x] The step size handling in CG/BFGS is somewhat problematic also because the line search step size is bounded to [1e-8, 50] and these bounds become too restrictive if the gradients are very large/small. Something could be done to this.
- [x] Maybe test the initial step is actually close to norm 1

See gh-3581 and https://stackoverflow.com/questions/33853929/fmin-cg-desired-error-not-necessarily-achieved-due-to-precision-loss for a problem this causes

### CODE REVIEW

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -651,6 +652,56 @@ def dY_dx(x, c=None):
...
+            assert_(res.success, err_msg)
+            assert_allclose(res.x, [1.0], err_msg=err_msg)
+            assert_(res.nit <= 3, err_msg)
+
+            if scale > 1e-10:
+                if method in ('CG', 'BFGS'):
+                    assert_allclose(first_step_size[0], 1.01, err_msg=err_msg)
+                else:
+                    assert_(first_step_size[0] > 0.5 and first_step_size[0] < 3,
+                            err_msg)
```
Comment (dlax) [inline scipy/optimize/tests/test_optimize.py]: Could you add a comment explaining why you check for an interval here? Also about the values maybe.

**Comment by dlax** [comment]
Looks good to me overall.

File: scipy/optimize/tests/test_optimize.py (hunk 1 above)
Comment (pv) [inline scipy/optimize/tests/test_optimize.py]: Ok. The changes I made only affect BFGS and CG. LBFGS and Newton-CG have slightly different logic, but are scaling invariant.

**Comment by pv** [comment]
Updated.

File: scipy/optimize/tests/test_optimize.py (hunk 1 above)
Comment (dlax) [inline scipy/optimize/tests/test_optimize.py]: Pauli Virtanen wrote:
> Ok. The changes I made only affect BFGS and CG. LBFGS and Newton-CG have
> slightly different logic, but are scaling invariant.
>

Ok, makes sense, thanks!

**Pull request merged**